Readme.txt
0. Licence
---
This software is distributed under the MIT licence.


1. Introduction
---

These programs (Pepl) implement and demonstrate parameter estimation
(PE) in SLPs, using the FAM algorithm. For more information on how
to use these programs see, doc/user_guide.ps.gz. In doc/pfam.ps.gz
are some reading notes on the implementation of the FAM algorithm.

2. Installation
---

You most likely already done that. Pe-pl is distributed as Prolog
source code that can be run on the following Prolog systems:

	Yap (tested on 6.3)
			http://www.dcc.fc.up.pt/~vsc/Yap
	Swi tested on 7.1.4.
			http://www.swi-prolog.org/

   Older versions are known to run on: 

	SICStus (last tested on 3.10, Pepl sources are now well out of date)
			http://www.sics.se/isl/sicstus.html

	On Swi you can install with 
		?- pack_install( pepl ).

	On Yap simply download the latest sources from
		http://stoics.org.uk/~nicos/sware/pepl
	then gunzip and untar the downloaded file.

3. Quick start.
---

On SWI, start swipl, and

[library(pepl)].

[pack('pepl/examples/main')]

main.
main_store.
main_sample.
main_exact. (alias for main)

On Yap, cd to the Pepl's examples directory start yap, and then

[main].
main.
main_store.
main_sample.
main_exact. (alias for main)

See main_* files in examples/ for more examples and doc/ for documentation.

4. Contact
---

Nicos Angelopoulos, http://stoics.org.uk/~nicos

London, February 2017


doc/Releases.txt
2.4
    @ 2023/05/08     DOC:    fixed typo

2.3                  @ 2023/05/06
    @ 2023/05/05     IFACE:  sample/1, sample/5, in new file src/sample.pl
                     IFACE:  seed_pe/0 in new file src/seed_pe.pl
                     SLP:    slp/doubles.slp, demos dead probability
                     IFACE:  scall_sum/2, scall_findall/2, allow X-Goal in 1st argument
                     CHANGE: scall/1, this is now non-sampling
                     IFACE:  scall/2, adds probability of derivation 
                     CHANGE: scall_1/6 -> resolution_pe/6, now in src/resolution_pe.pl
                     CHANGE: scall/6 -> scall/5, completes separation of scall_?/N and sample/M preds
                     CHANGE: hide some of the priate loads to init_lib
                     DOC:    inference on module docs

2.2                  @ 2022/01/02
    @ 2022/01/02     CHANGED  reporting of loading via print_message/2
    @ 2022/01/01     DOC      of scall/1 and sprucing up scall/6
                     EXAMPLE  Non failure version of member3.slp in member3_nf.slp
                     DOC      format options of fam/1

2.1 @ 2017/02/25     Added license (MIT) and uploaded to github.
        -            Added pepl_citation/2, with the PLP paper details.
                     Version numberss from now on will follow the 2 part 
							stoics convention.
							Fixed broken translation of A->B;C in sources (particularly non stochastic clauses).
							Added simple coin slp.

2.0.6   2014/01/28   Minor corrections to docs and website.
                     Changed operator's :: priority to 600 for compatible with 
							logtalk. Needs to be confirmed by Paulo Moura.
							Some corrections to examples/turtles/.

2.0.4   2014/01/27   2.0.2 did n't have all required files 
                     (sym links in src/ were all missing).    Paulo Moura

2.0.2   2014/01/26   removed email from Readme.txt which showed on Swi website.

2.0.0   2014/01/26   publish as Swi pack. Also tested on dev. Yap

1.0.0   ???          first public version for Yap and Sicstus.


src/pepl.pl
/** <module> An implementation of the FAM algorithm.

Pepl is an implemention of the failure adjusted (FAM) algorithm which does 
parameter estimation (PE) of the probability labels of stochastic logic programs (SLPs).

See documentation fam/1 for details on how to run parameter estimation on SLPs.

Example stochastic programs are in directory =|slp|= and example run scripts are in =|examples|=.

Licence
---
This software is distributed under the MIT licence.

## Installation and testing ...

Pepl runs on current versions of SWI (7) and Yap (6.3).

### ... on SWI

==
pack_install(pepl).
[library(pepl)].
[pack('pepl/examples/main')].
main.
==

### ... on Yap

Download latest sources from http://stoics.org.uk/~nicos/sware/pepl
or https://github.com/nicos-angelopoulos/pepl

==
gunzip pepl-*tgz
tar xf pepl-*tar
cd pepl-*
cd examples
yap
[main].
main.
==

## Resolution
In addition to parameter estimation Pepl implements two way of performing resolution over SLPs:
stochastic sampling resolution and SLD-based probabilisic inference.

### Stochastic sampling resolution

These predicates allow to sample from a loaded stochastic logic program (Slp). The resolution strategy here
are that of chosing between probabilistic choices according to their relative values. The main idea is that
when sampling many times from a top goal will in the long run sample each derivation path in proportion 
to the probability of the derivation. The probability of a derivation/refutation, is simply the product 
of all the probabilities attached to resolution steps during the derivation. 


For very deep probabilistic programs, it is sometimes useful to but a minimum value of probability
we are interested in. This is a way to shorten the search space without losing significant amounts
of probability mass (see second argument of sample/5).

See
  * sample/1
  * sample/5

### SLD-based probabilisic inference

These predicates allow standard SLD exploration of a stochastic query against an SLP. Predicates here
allow to explore what is derivable and often attach a probability and ather information to each derivation.

Note that in probabilistic inference we often are more interested in failures than in standard LP. 
This is because there is a probability mass loss which each failed probabilistic branch.

Probabilistic inference predicates
  * scall/1
  * scall/2
  * scall/5
  * scall_findall/2
  * scall_sum/2

## Predicates index

  * fam/1
  * sample/1, sample/5
  * scall/1, scall/2, scall/5
  * scall_findall/2, scall_sum/2
  * sload_pe/1, sload_pe/2
  * ssave/1, sls/0
  * seed_pe/0
  * switch_dbg/1, dbg_pepl/1
  * pepl_citation/2, pepl_version/2

## Package information

@author Nicos Angelopoulos
@license This software is distributed under the MIT licence
@version 2.3, 2023/5/6, added extensive sampling and inference support
@version 2.2, 2022/1/2
@version 2.1, 2017/2/25
@version 2.0.6, 2014/01/28 
@see  the user guide at pack('pepl/doc/pepl-user_guide.pdf').
@see James Cussens. Parameter estimation in stochastic logic programs. Machine Learning, 44(3):245-271, 2001. ftp://ftp.cs.york.ac.uk/pub/aig/Papers/james.cussens/jcslpmlj.pdf
@see Nicos Angelopoulos, Notes on the implementation of FAM, 3rd Probabilistic Logic Programming workshop (a ILP 2016 workshop), 03/09/2016, http://ceur-ws.org/Vol-1661/paper-05.pdf
@see pepl website http://stoics.org.uk/~nicos/sware/pepl
*/

%% pepl_citation( -Atom, -Bibterm ).
%
% This predicate succeeds once for each publication related to this library.
% Atom is the atom representation
% suitable for printing while Bibterm is a bibtex(Type,Key,Pairs) term of the same publication.
% Produces all related publications on backtracking.
%
pepl_citation( Atom, bibtex(Type,Key,Pairs) ) :-
     Atom = 'Notes on the implementation of FAM\nNicos Angelopoulos\n3rd Probabilistic Logic Programming workshop (PLP 2016, a workshop of ILP 2016). September 2016, Imperial College London. Pages 46-58.',
    Type = inproceedings,
    Key  = 'AngelopoulosN+2016',
    Pairs = [
               author = 'Nicos Angelopoulos',
               title  = 'Notes on the implementation of FAM',
               booktitle = '3rd Probabilistic Logic Programming Workshop (collocated with ILP 2016)',
               year = 2016,
               month = 'September',
               address = 'Imperial College, London',
               publisher = 'CEUR',
               volume   = '1661',
               url     = 'http://ceur-ws.org/Vol-1661/'
     ].



/** pepl_version( -Version, -Date ).

Pepl's current Version (Maj:Min:Fix) and publication date (date(Year,Month,Day)).

==
?- pepl_version(V,D).
V = 2:3:0,
D = date(2021, 5, 6).
==

@version 2:3:0 2023/05/06
@version 2:2:0 2021/01/01

*/
pepl_version( 2:3:0, date(2023,5,6) ).
